text/0000-temp-components-lib.md

```diff
@@ -0,0 +1,609 @@
+3. High performance React patterns
+
+Essentially, this library should fulfill the Pando core value of creating a tool
+that helps its users become better Javascript and React developers.
+
+## Basic example
+
+For a basic components, like a Button, Field Message, or Input, it should function
+something like the following example.
+
+Here is an example of a form:
+
+```typescript
+import {
+  Button,
+  ErrorMessage,
+  Input,
+  Label,
+  FieldMessage,
+  FieldProvider
+} from '@pluralsight/react'
+
+...
+
+<form onSubmit={handleSubmitForm}>
+  <FieldProvider
+    id="name"
+    invalid={testValidity}
+    messageId="[nameMessage, nameErrorMessage]"
+    required
+  >
+    <Label>Name</Label>
+    <Input
+      placeholder="Enter your name"
+      type="text"
+      value={value}
+    />
+    <FieldMessage>Enter your full name.</FieldMessage>
+    <ErrorMessage>Please enter your full name.</ErrorMessage>
+  </FieldProvider>
+
+  <Button disabled={formNotValid} sentiment="action" type="submit">
+    Submit
+  </Button>
+</form>
+```
+
+Each component should be composable (can be used like Lego blocks) to allow maximum
+flexibility in creating "custom designs" that either include/remove specific elements.
+
+**This is the current standard supported and promoted by the React team.**
+
+In this example, the `FieldProvider` would act as the `getFormControlProps`
+Headless-styles helper in managing the state of the field and other children.
+
+This means, any field related component (i.e. Label, Input, ErrorMessage, etc.)
+would naturally inherit the state and all a11y props needed that is managed from
+the Headless-styles helpers (invalid, disabled, readOnly, etc.) and require no
+props of those from the user.
+
+This would also allow us to internally create error logs that would "make it hard
+to break something" which would inform the user when they are missing a FieldProvider.
+
+For a more complex component - see the [Detailed design](#detailed-design) section.
+
+## Motivation
+
+### Why are we doing this?
+
+There has been consistent requests for a React component library due to the nature
+of most developers either being either unaware of how to use a "headless" library
+or generally uncomfortable with being responsible for owning their own components.
+
+Having a component library that has a familiar interface which ultimately uses
+Headless-styles and React-utils under the hood should help break that barrier.
+
+### What use cases does it support?
+
+A component library would be most beneficial to teams/users that:
+
+1. Are mostly Backend developers that have little to no experience using Javascript or React.
+2. Junior frontend developers who have little to no experience in Javascript or React.
+3. Marketing departments who would like to use Pando, but have "just enough experience to be dangerous" in frontend development.
+4. Team leads that haven't been able to keep up with the technology trends but still need to code when possible.
```

Comment:
I think what another meaningful example here would be teams that are fully capable and perfectly competent with JS & React, but just aren't able to prioritize spending the time to hand-wire all the components together because their team needs to spend more of their time elsewhere. I know that if we didn't already have a component library my team uses it would be a daunting task to integrate and test using HS directly, and I'm very experienced with React/JS, if only due to the time cost of manually wiring things up and testing everything. It's not inherently _hard_, it's just a lot of additional effort that wasn't previously required because you're used to using an abstraction layer (i.e. components), and that additional time can add up very quickly. When faced with the choice to wire up headless styles manually (in the general sense, not this library specifically) or implement a new feature, I think most teams would prefer to get the new feature built, especially when it's tied to OKRs.

I would guess that that's probably the biggest reason for hesitancy with adopting HS directly and the main source of friction: it's not that teams are unable to or that they're not experienced enough to implement it, it would just take more time that they don't have, and it's harder to prioritize introducing a whole new paradigm into the way code is written. The cost of change is much higher that way.

Reply:
Let me go ahead and add that point since it is extremely valid.

Reply:
Done.